### PR TITLE
Multiple arguments in interface

### LINKED ISSUE #14: Multiple arguments
How do we include methods in the interface that take several arguments of different types?

### PR DESCRIPTION
**Pull request by gdalle**
Fix #14.

- [x] Define an `Arguments` type wrapping a `NamedTuple`
- [x] Add an automatic check to `Interfaces.test` verifying that each test object is either
  - of the given type `O`
  - of type `Argument` with the given type `O` for its first field
- [x] Write a test suite for multiple arguments (the `GroupInterface`)
- [x] Write documentation

### CODE REVIEW

File: src/arguments.jl [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+for f in (:length, :firstindex, :lastindex, :isempty, :keys, :values)
+    @eval Base.$f(a::Arguments) = $f(nt(a))
+end
+
+for f in (:(==), :isequal, :<, :isless, :isapprox)
+    @eval Base.$f(a1::Arguments, a2::Arguments) = $f(nt(a1), nt(a2))
+end
+
+Base.iterate(a::Arguments, iter=1) = iterate(nt(a), iter)
```
Comment (rafaqz) [inline src/arguments.jl:26]:
```suggestion
Base.iterate(a::Arguments, iter...) = iterate(nt(a), iter...)
```
The default argument isn't needed

File: src/arguments.jl [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+struct Arguments{names,T}
+    nt::NamedTuple{names,T}
+end
+
+Arguments(; kw...) = Arguments(values(kw))
+Arguments{names}(vals) where {names} = Arguments(NamedTuple{names}(vals))
+
+nt(a::Arguments) = getfield(a, :nt)
+
+for f in (:length, :firstindex, :lastindex, :isempty, :keys, :values)
```
Comment (rafaqz) [inline src/arguments.jl:18]: Do we need to define `propertynames` as well?

File: src/test.jl [hunk 3]
```diff
@@ -8,6 +8,15 @@ Base.iterate(tow::TestObjectWrapper, args...) = iterate(tow.objects, args...)
 Base.length(tow::TestObjectWrapper, args...) = length(tow.objects)
 Base.getindex(tow::TestObjectWrapper, i::Int) = getindex(tow.objects, i)
 
+function check_coherent_types(O::Type, tow::TestObjectWrapper)
+    for obj in tow
+        coherent_types = obj isa O || (obj isa Arguments && first_field_type(typeof(obj)) <: O)
+        if !coherent_types
+            throw(ArgumentError("Each tested object must either be an instance of $O or an instance of Arguments whose first field type is $O"))
```
Comment (rafaqz) [inline src/test.jl:15]:
```suggestion
            throw(ArgumentError("Each tested object must either be an instance of `$O` or an instance of `Arguments` whose first field type is `$O`"))
```

**Review by rafaqz — CHANGES_REQUESTED** [review]
Nice! That came together quickly.

Only a few minor comments.

File: test/groupinterface.jl [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            "inversion works" => a::Arguments -> multiplication(a.x, inversion(a.x)) ≈ neutral(typeof(a.x)),
+        ),
+    ),
+    optional = (;
+        abelian = (
+            "multiplication commutes" => a::Arguments -> multiplication(a.x, a.y) ≈ multiplication(a.y, a.x),
+        ),
+    )
+) """
+A group is a set of elements with a neutral element where you can perform multiplications and inversions.
```
Comment (rafaqz) [inline test/groupinterface.jl:31]: It would be nice to demonstrate documenting the required `Arguments` here

Comment (gdalle) [reply test/groupinterface.jl:31]: Yeah, I just would like to wait until the docs website is back and running, to set up a Literate pipeline that removes the need for copy paste. What do you think?

Comment (gdalle) [reply test/groupinterface.jl:31]: Basically that way we could put the docs in the test file

Comment (rafaqz) [reply test/groupinterface.jl:31]: Done. Looks like I just added the DOCUMENTER_KEY a year ago and forgot about the deploy keys :sweat_smile:

Comment (rafaqz) [reply test/groupinterface.jl:31]: I havent used Literate.jl like that so interested to see what you do.

Comment (gdalle) [reply test/groupinterface.jl:31]: I'll do it in a separate PR for the existing code first, to avoid bloating this one

Comment (gdalle) [reply test/groupinterface.jl:31]: #17

File: src/arguments.jl (hunk 2 above)
Comment (gdalle) [reply src/arguments.jl:18]: doesn't cost much

File: src/arguments.jl (hunk 1 above)
Comment (gdalle) [reply src/arguments.jl:26]: okay

File: src/test.jl (hunk 3 above)
Comment (gdalle) [reply src/test.jl:15]: done

**Comment by rafaqz** [comment]
But if we dont check the type what proof do qe have that the test object is the same as the type we are declaring implements the interface?

Or does thst check happen elsewhere?

**Comment by rafaqz** [comment]
There has to be a little policing... so that passing the tests means something to other users.

I originally had this all locked down so there was no way the traits could work without the correct tests passing. Running tests in precompilstion can guarantee this and I think we should remain open to returning yo that stratedgy.

Maybe we need a standard argument name for the target object rather than the first, then the tesrs are easier to read.

**Comment by gdalle** [comment]
But in the group scenario for instance there are several target objects

**Comment by gdalle** [comment]
Maybe Arguments{T} where T is the interface-implementing type? And then we check in the background that at least one field of args subtypes T?

**Comment by gdalle** [comment]
That way we can test at argument construction

**Comment by rafaqz** [comment]
Although that will make the implementer soe ify the type in Arguments. But we already have it available in `test` so we can just check all argument types there?

Im  still wondering if maybe we can get the correct argument name from `@implements` somehow so its totally locked down.

But maybe we dont have to get it perfect straight away.

**Comment by gdalle** [comment]
How about:
1. The instances of the interface-implementing type must have a name starting with obj (which allows obj1 and obj2)
2. There must be at least one such instance in the fields of an Arguments

**Comment by rafaqz** [comment]
Yes thst could work. Lets sit on this for a while and just tests that any one argument matches the type in `test` fir now

**Comment by gdalle** [comment]
Ok I'll write that. Just a silly question: given a tuple type `Tuple{Int,Float64,String}`, how would you list the field types?

**Comment by rafaqz** [comment]
`fieldtypes(T)`?

Or just map over the values with 

```julia
any(map(x -> x isa T, nt(args)))
```

**Comment by gdalle** [comment]
Yeah I wanted to deduce it from the type itself but fieldtypes didn't work, just wondered if you had encountered that before. I'll check the values instead

**Comment by gdalle** [comment]
@rafaqz this should be good to go as a first draft, even if we end up refining it later

**Comment by rafaqz** [comment]
Ok looks like it just needs a rebase

**Comment by gdalle** [comment]
done

**Review by rafaqz — COMMENTED** [review]
LGTM

**Pull request merged**